Jenkinsfile

```diff
@@ -0,0 +1,188 @@
+      //write payload to a file
+      writeFile file: "/tmp/packer_pipeline_job_build_$BUILD_NUMBER", text: "$params.payload"
+    
+      //pass this output to our script
+      env.payload_parsed_JSON = sh(returnStdout: true, script: """ 
+        export PACKER_TEMPLATES_DIR=$WORKSPACE
+        cat /tmp/packer_pipeline_job_build_$BUILD_NUMBER | $WORKSPACE/files/default/bin/packer_pipeline.rb
+      """).trim()
+      writeFile file: "/tmp/${JOB_NAME}-${BUILD_NUMBER}.json", text: env.payload_parsed_JSON
+   }   
+
+   /* actually starts processing the templates on the right nodes */
+   stage('start_processing_on_right_nodes') {
+
+      env.pr = get_from_payload('pr')
+      
+
+      //TODO: this should be set from the job
+      //set path to the packer binary
+      env.packer = '/usr/local/bin/packer'
+      env.PATH = "/usr/libexec:/usr/local/bin:/opt/chef/embedded/bin:${env.PATH}"
+
+      //TODO: this should be set from the job as an env variable
+      //this should *ALWAYS* match what lib/packer_pipeline.rb return
+      archs = ['x86_64', 'ppc64']
+
+      //TODO: parallelize this -- both archs can be processed seperately after all!
+      for ( arch in archs ) {
+         env.arch = arch
+         templates = get_from_payload(env.arch)
+         if ( ! templates.empty && templates != null ) {
+             echo "Starting execution for $env.arch"
+             //do following things on the node.
+             node (env.arch) {
+               clone_repo_and_checkout_pr_branch()
+               run_linter(env.arch)
+               build_image(env.arch)
+               deploy_image_for_testing(env.arch)
+               run_tests(env.arch)
+               archive '*'
+               //deploy_on_production() -- seperate!
+             }
+         }
+         else
+         {
+            echo "No templates for $env.arch!"
+         }
+      }
+   }
+}
+
+/*
+get_from_payload(variable_name) :
+   read the parsed payload JSON from the disk
+   and return variables in a form that's easy to process
+
+   this always runs on the master.
+
+Why do we need this function?
+-----------------------------
+
+Jenkins requires any variable to be serializable so that
+even if the master is restarted, the state is maintained.
+A JSON object in memory is represented as a Map which does not
+implement the Serializable interface.
+
+So we implement our own "serializer-deserializer" in this manner.
+*/
+def get_from_payload(v) {
+   def jsonSlurper = new JsonSlurper()
+   def data = jsonSlurper.parseText("${payload_parsed_JSON}")
+
+   if ( data.containsKey(v) ) {
+      r = data[v]
+   }
+   else {
+      r = null
+   }
+   println "Returning $r"
+   return r
+}
+
+
+// clone the packer-templates repo and checkout the branch which we want to use
+def clone_repo_and_checkout_pr_branch() {
+   stage('clone_repo_and_checkout_pr_branch') {
+       //checkout all templates
+       git 'https://github.com/osuosl/packer-templates'
+          sh "git pr $env.pr"
```

Comment:
This is defined at https://github.com/osuosl-cookbooks/osl-jenkins/blob/master/files/default/gitconfig#L11